sections/advanced/reaching-ui-state.md

```diff
@@ -0,0 +1,222 @@
+      const data = response.response.body;
+      // We push a new entry into the xhrData array
+      xhrData.push({ url, method, data });
+    }
+  });
+  // this is used as a ‘filter’ for the data you want to record.
+  // You can specify the methods and routes that are recorded
+  cy.log('recording!');
+  cy.route({
+    method: 'GET',
+    url: '*',
+  });
+  // if recording, after the test runs, create a fixture file with the recorded data
+  after(function () {
+    cy.writeFile(`./cypress/fixtures/${pathToJson}.json`, xhrData);
+    cy.log(`Wrote ${xhrData.length} XHR responses to local file ${pathToJson}.json`);
+  });
+}
+
+/** Plays recorded fixture with all required network data as json*/
+function playStubbedFixture(stateFixture) {
+  cy.log(`playing fixture from ${stateFixture}`);
+  cy.fixture(stateFixture, { timeout: 15000 }) // the fixture file may be large and take time in CI
+    .then(data => {
+      for (let i = 0, length = data.length; i < length; i++) {
+        cy.route(data[i].method, data[i].url, data[i].data);
+      }
+    }).as(`stateFixture_stub`);
+}
+
+/** Visits the stubbed state */
+function visitStubbedState(stubFile, url, wait: boolean = true) {
+  playStubbedFixture(stubFile);
+  cy.visit(url);
+  if (wait) { // sometimes you do not want to wait for network, this gives you the option
+    cy.wait('@stateFixture_stub', { timeout: 15000 });
+  }
+}
+
+//////////
+// usage
+
+// recording network
+it('should run your test', function () {
+  stubrecorder('jsonfileNameForNetworkData');
+
+  // your original test
+
+  cy.wait(5000); // one-time wait so that the after() step records all the network without missing anything
+
+  // the rest of your original test
+});
+
+// playing the stubbed network
+it('should run your test', function () {
+  // every time we visit this endpoint, all network will be stubbed
+  // double check this by observing (XHR stubbed) network responses in the test runner
+  visitStubbedState('jsonfileNameForNetworkData', '/endpoint');
+
+  // the rest of your original test
+});
+```
+
+
+<br/><br/>
+
+### Application actions
+
+Cypress gives complete control of your application. You can bypass the page object abstraction layer (which is detached from your application) access the UI directly via `cy.get()` , have access to the API and database, and even access the source code.
+
+Component testing is an experimental feature in Cypress as of version 4.7.0, albeit some of the functionality is available. Refer to component testing articles under tools section for experiments with React and Storybook on this topic.
+
+![devtools > network tab](../../assets/images/ui-state/app-action.PNG)
+
+
+Here is a quick example of how you would allow source code access to Cypress in an Angular application.
+```javascript
+// Angular Component file example
+/* setup:
+ 0. Identify the component in the DOM;
+  inspect and find the corresponding <app.. tag,
+  then find the component in src
```

Comment:
A simpler `Right in the constructor of your component...` could be better, IMHO

Reply:
modified